sycl/doc/design/SYCLBINDesign.md

```diff
@@ -0,0 +1,280 @@
+| `uint32_t` | Magic number. (0x53594249)                                                    |
+| `uint32_t` | SYCLBIN version number.                                                       |
+| `uint32_t` | Number of abstract modules.                                                   |
+| `uint32_t` | Number of IR modules.                                                         |
+| `uint32_t` | Number of native device code images.                                          |
+| `uint64_t` | Byte size of the metadata byte table.                                         |
+| `uint64_t` | Byte size of the binary byte table.                                           |
+| `uint64_t` | Byte offset of the global metadata from the start of the metadata byte table. |
+| `uint64_t` | Byte size of the global metadata from the start of the metadata byte table.   |
+
+
+#### Global metadata
+
+The global metadata entry contains a single property set with the identifying
+name "SYCLBIN/global metadata", as described in the
+[PropertySets.md](PropertySets.md#syclbinglobal-metadata) design document.
+
+
+### Abstract module
+
+An abstract module is a collection of device binaries that share properties,
+including, but not limited to: kernel names, imported symbols, exported symbols,
+aspect requirements, and specialization constants.
+
+The device binaries contained inside an abstract module must either be an IR
+module or a native device code image. IR modules contain device binaries in some
+known intermediate representation, such as SPIR-V, while the native device code
+images can be an architecture-specific binary format. There is no requirement
+that all device binaries in an abstract module are usable on the same device or
+are specific to a single vendor.
+
+
+#### Abstract module header
+
+A abstract module header contains the following fields in the stated order:
+
+| Type       | Description                                                                                |
+| ---------- | ------------------------------------------------------------------------------------------ |
+| `uint64_t` | Byte offset of the metadata from the start of the metadata byte table.                     |
+| `uint64_t` | Byte size of the metadata in the metadata byte table.                                      |
+| `uint32_t` | Number of IR modules.                                                                      |
+| `uint32_t` | Index of the first IR module header in the IR module header array.                         |
+| `uint32_t` | Number of native device code images.                                                       |
+| `uint32_t` | Index of the first native device code images header native device code image header array. |
+
+#### Abstract module metadata
+
+An abstract module metadata entry contains any number of property sets, as
+described in [PropertySets.md](PropertySets.md), excluding:
```

Comment:
Not sure if this is a problem, but what if we add a new property there which cannot be applied to an abstract module?

Reply:
I don't think adding new cases should be a problem. It is mostly to say that these property sets have no meaning for the abstract module metadata block. I don't think it's worth checking whether they are there or not, over just skipping them.